SourceCode/raytracer/Triangle.h

```diff
@@ -0,0 +1,160 @@
+#pragma once
+#include "CRTTypes.h"
+
+// hw4 Task 1
+class Triangle {
+public:
+    Vec3 v0, v1, v2;
+
+    Triangle() : v0(Vec3()), v1(Vec3()), v2(Vec3()) {}
+
+    Triangle(const Vec3& v0, const Vec3& v1, const Vec3& v2) : v0(v0), v1(v1), v2(v2) {}
+
+    Vec3 normal() const {
+        Vec3 e1 = v1 - v0;
+        Vec3 e2 = v2 - v0;
+        return e1.cross(e2).normalize();
+    }
+
+    float area() const {
+        Vec3 e1 = v1 - v0;
+        Vec3 e2 = v2 - v0;
+        return e1.cross(e2).length() * 0.5f;
+    }
+
+    Triangle transform(const Matrix3x3& mat) const {
+        Vec3 newV0 = mat * v0;
+        Vec3 newV1 = mat * v1;
+        Vec3 newV2 = mat * v2;
+        return Triangle(newV0, newV1, newV2);
+    }
+
+    std::string toString() const {
+        return "Triangle: {" + v0.toString() + ", " + v1.toString() + ", " + v2.toString() + "}";
+    }
+
+    /*
+    * Input: ray, rayProj
+    * Input: rayProj: projection of ray.direction onto camera direction
+    * Output: t: distance from ray.origin to the intersection point
+    * Output: p: intersection point
+    * return: true if the ray intersects the triangle
+    */
+    bool intersect_simple(const Ray& ray, float& t, Vec3& p) const {
+        // Here assuming counter-clockwise order
+        Vec3 e0 = v1 - v0;
+        Vec3 e1 = v2 - v0;
+        Vec3 tri_normal = cross(e0, e1).normalize();
```

Comment:
You could cache the normal in the triangle instead of computing it each time you need it.

Reply:
@VladislavVulchevChaos  I agree. In a CUDA ray tracer, do you typically precompute the normal too, or is it better to compute it on the fly to improve memory coalescing?